languages/base.lua

```diff
@@ -0,0 +1,167 @@
+--- SILE language class.
+-- @interfaces languages
+
+local language = pl.class()
+language.type = "language"
+language._name = "base"
+
+local loadkit = require("loadkit")
+local setenv = require("rusile").setenv
+
+-- Allows loading FTL resources directly with require(). Guesses the locale based on SILE's default resource paths,
+-- otherwise if it can't guess it Loads assets directly into the *current* fluent bundle.
+local require_ftl = loadkit.make_loader("ftl", function (file)
+   local contents = assert(file:read("*a"))
+   file:close()
+   return assert(fluent:add_messages(contents))
+end)
+
+function language:_init (typesetter)
+   self.typesetter = typesetter
+   self:_declareBaseSettings()
+   self:declareSettings()
+   self:_registerBaseCommands()
+   self:registerCommands()
+   self:loadMessages()
+   self:setupNodeMaker()
+   self:setupHyphenator()
+end
+
+function language:_post_init ()
+   SILE.settings:registerHook("document.language", function (lang)
+      self.typesetter:switchLanguage(lang)
+   end)
+end
+
+-- TODO: reconsider naming of 'setup' and 'nodeMaker'
+function language:setupNodeMaker ()
+   -- TODO should this be an instance of a constructor? inconsistent with typesetter/language/class/etc.
+   self.nodeMaker = require("languages.base-nodemaker")
+end
+
+function language:setupHyphenator ()
+   -- TODO should this be an constructor instead of an instance? inconsistent with typesetter/language/class/etc.
+   self.hyphenator = require("languages.base-hyphenator")(self)
+end
+
+function language:activate ()
+   local lang = self:_getLegacyCode()
+   fluent:set_locale(lang)
+   os.setlocale(lang)
+   setenv("LANG", lang)
+end
+
+function language:_getLegacyCode ()
+   return self._name
+end
+
+-- TODO: not about hyphenation rules per warning....
```

Comment:
Lot's of TODO in this file, not all very clear. No idea for instance what this one is supposed to mean.

Reply:
Fair enough. This was a note that the message in `SU.warn()` needed updating to reflect the actual function, not mixed with notes about hyphenation support. The trouble is I fixed the task without removing the TODO note. (Also working through other TODOS over here...)